### PR TITLE
[SPARK-44987][SQL] Assign a name to the error class `_LEGACY_ERROR_TEMP_1100`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to assign the name `NON_FOLDABLE_ARGUMENT` to the legacy error class `_LEGACY_ERROR_TEMP_1100`, and improve the error message format: make it less restrictive.

### Why are the changes needed?
1. To don't confuse users by slightly restrictive error message about literals.
2. To assign proper name as a part of activity in SPARK-37935

### Does this PR introduce _any_ user-facing change?
No. Only if user's code depends on error class name and message parameters.

### How was this patch tested?
By running the modified and affected tests:
```
$ build/sbt "test:testOnly *.StringFunctionsSuite"
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
$ build/sbt "core/testOnly *SparkThrowableSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -2210,6 +2210,12 @@
     ],
     "sqlState" : "42607"
   },
+  "NON_FOLDABLE_ARGUMENT" : {
+    "message" : [
+      "The function <funcName> requires the parameter <paramName> to be a foldable expression of the type <paramType>, but the actual argument is a non-foldable."
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:2215]: How would you improve the error message?

**Comment by MaxGekk** [comment]
@dongjoon-hyun @HyukjinKwon Could you review this PR, please.

**Comment by dongjoon-hyun** [comment]
Sure, @MaxGekk .

File: common/utils/src/main/resources/error/error-classes.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply common/utils/src/main/resources/error/error-classes.json:2215]: Ya, in this case, new message looks reasonable to me. :)

File: docs/sql-error-conditions.md [hunk 2]
```diff
@@ -1299,6 +1299,12 @@ Cannot call function `<functionName>` because named argument references are not
 
 It is not allowed to use an aggregate function in the argument of another aggregate function. Please use the inner aggregate function in a sub-query.
 
+### NON_FOLDABLE_ARGUMENT
+
+[SQLSTATE: 22024](sql-error-conditions-sqlstates.html#class-22-data-exception)
```
Comment (dongjoon-hyun) [inline docs/sql-error-conditions.md:1304]: Data exception category is also reasonable to me if there is no other proper one.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @MaxGekk .

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @dongjoon-hyun for review.

**Pull request closed without merging**